[Feature] Support CUME_DIST window function

Moonm3n opened the pull request:
## What type of PR is this:
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [x] UT
- [x] Doc
- [ ] Tool

## Which issues of this PR fixes:
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #13021 #19632

## Problem Summary(Required):
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

The CUME_DIST window function returns the cumulative distribution of a value within a group of values; that is, the percentage of partition values less than or equal to the value in the current row. This represents the number of rows preceding or peer with the current row in the window ordering of the window partition divided by the total number of rows in the window partition. Return values range from 0 to 1.

It can not been used with the windowing clause. And for the implementation, it uses `RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW`.

NOTE that, CUME_DIST need the boundary of partition to calculate its value. For these functions, we must wait util the partition finished. Therefore, it uses `_process_by_partition_for_unbounded_preceding_range_frame_materializing`.

The CUME_DIST window function is only supported by the pipeline engine. BE will report the NotSupported error, when non-pipeline engine executes CUME_DIST in BE.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
- [x] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

Moonm3n committed:
[Feature] Support CUME_DIST window function

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=22775) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=22775) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=22775) it.</sub>

Moonm3n committed:
[Doc] Add document of CUME_DIST window function

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

Moonm3n committed:
[Refactor] Fix clang-format style issues

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

Moonm3n committed:
[Test] Add fe test of CUME_DIST window function

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[336]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

Moonm3n committed:
[Test] Add SQL test of CUME_DIST window function

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

fzhedu commented on docs/sql-reference/sql-functions/Window_function.md:
```diff
@@ -196,6 +198,44 @@ from int_t where property in ('odd','even');
+### CUME_DIST()
+
+The CUME_DIST() function calculates the cumulative distribution of a value within a partition, indicating its relative position as a percentage of values less than or equal to the value in the current row. With a range of 0 to 1, it's useful for percentile calculations and data distribution analysis.
+
```
please add some notes for the case without order by.
```
**This function should be used with ORDER BY to sort partition rows into the desired order. Without ORDER BY, all rows are peers and have value N/N = 1, where N is the partition size.**
```

Moonm3n replied:
Done.

fzhedu commented on test/sql/test_window_function/R/test_cume_window_function:
```diff
@@ -0,0 +1,705 @@
+       (7, 3, 60, 14.5, '2023-01-28', 'dewberry');
+-- result:
+-- !result
+SELECT *, CUME_DIST() OVER (ORDER BY v1) AS cume_dist_v1 FROM cume_test;
```
please add some cases without order-by clause, and the data with null values.

Moonm3n replied:
Done.

Moonm3n committed:
[Test] Fix the TOP-N push down test of CUME_DIST window function

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[336]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

Moonm3n committed:
[Doc] Update doc of CUME_DIST window function

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

Moonm3n committed:
[Test] Update SQL test of CUME_DIST window function
1. add test with null value.
2. add test without order by.

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[336]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

Moonm3n committed:
[Test] Update test of CUME_DIST window function

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[336]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[336]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

fzhedu reviewed (dismissed):
LGTM

Moonm3n committed:
[Test] Update test of CUME_DIST window function

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[356]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

Moonm3n committed:
fix sql test

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[356]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

liuyehcf commented on be/src/exec/pipeline/analysis/analytic_sink_operator.cpp:
```diff
@@ -59,6 +68,7 @@ Status AnalyticSinkOperator::prepare(RuntimeState* state) {
                         &AnalyticSinkOperator::_process_by_partition_for_unbounded_preceding_rows_frame_materializing;
             }
         } else {
+            // ROWS BETWEEN N PRECEDING AND CURRENT ROW
```
N PRECEDING AND M FOLLOWING 
or
N PRECEDING AND CURRENT ROW

satanson commented on be/src/exprs/agg/window.h:
```diff
@@ -233,6 +233,44 @@ class DenseRankWindowFunction final : public WindowFunction<DenseRankState> {
+    int64_t count;
+};
+
+class CumeDistWindowFunction final : public WindowFunction<CumeDistState> {
```
it seems that cume_dist show should only support range-typed frame, so we must fobid rows-typed frame for cume_dist

Moonm3n replied:
In this PR, the use of Windowing clause is not allowed with the `CUME_DIST()` window function, just like `Rank()` and `DENSE_Rank()`. I've forbidden it at the frontend.

```java
if (analyticExpr.getWindow() != null) {
    if ((isRankingFn(analyticFunction.getFn()) || isCumeFn(analyticFunction.getFn()) || 
            isOffsetFn(analyticFunction.getFn()) || isHllAggFn(analyticFunction.getFn()))) {
        throw new SemanticException("Windowing clause not allowed with '" + analyticFunction.toSql() + "'",
                analyticExpr.getPos());
    }

    verifyWindowFrame(analyticExpr);
}
```

Moonm3n commented on test/sql/test_window_function/R/test_cume_window_function:
```diff
@@ -0,0 +1,1708 @@
+8	6	None	None	2023-02-01	fig	1.0
+8	2	None	2.5	2023-01-29	None	1.0
+-- !result
+SELECT *, CUME_DIST() OVER () AS cume_dist_without_order_by FROM cume_test ORDER BY v2;
```
Yes, the order-by-clause is essential. Without ORDER BY, all rows are peers and have a value of N/N = 1, where  N represents the partition size.

Moonm3n commented on be/src/exec/pipeline/analysis/analytic_sink_operator.cpp:
```diff
@@ -273,6 +283,37 @@ void AnalyticSinkOperator::_process_by_partition_for_unbounded_preceding_rows_fr
+    if (_analytor->has_cume_function()) {
+        _analytor->set_partition_size_for_cume_function();
+    }
+    do {
```
I'm somewhat perplexed. According to the function `_process_by_partition_if_necessary_materializing`, computations are only performed when all data within a partition have arrived. Additionally, when calculating rank, I'm using `state.count = _partition_end - _partition_start;` instead of `chunk_size`. Would this approach potentially be accurate?

```c++
Status AnalyticSinkOperator::_process_by_partition_if_necessary_materializing() {
    while (_analytor->has_output()) {
        ...
        _analytor->find_partition_end();
        // Only process after all the data in a partition is reached
        if (!_analytor->found_partition_end().first) {
            return Status::OK();
        }

        const auto chunk_size = static_cast<int64_t>(_analytor->current_chunk_size());
        _analytor->create_agg_result_columns(chunk_size);

        bool is_new_partition = _analytor->is_new_partition();
        if (is_new_partition) {
            _analytor->reset_state_for_cur_partition();
        }

        (this->*_process_by_partition)(chunk_size, is_new_partition);

        // Chunk may contains multiply partitions, so the chunk need to be reprocessed
        if (_analytor->is_current_chunk_finished_eval(chunk_size)) {
            ChunkPtr chunk;
            RETURN_IF_ERROR(_analytor->output_result_chunk(&chunk));
            _analytor->offer_chunk_to_buffer(chunk);
        }
    }
    return Status::OK();
}
```

Moonm3n commented on be/src/exec/pipeline/analysis/analytic_sink_operator.cpp:
```diff
@@ -273,6 +283,37 @@ void AnalyticSinkOperator::_process_by_partition_for_unbounded_preceding_rows_fr
+    }
+    do {
+        _analytor->find_peer_group_end();
+        DCHECK(!_analytor->found_partition_end().first);
```
This line is unnecessary and was added during debugging. I will remove it.

Moonm3n committed:
[Enhancement] Remove useless code, add comment

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[356]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

Moonm3n commented:
This pull request has encountered some issues. Earlier, it was unable to pass the CI checks, and based on suggestions from the StarRocks team, I rebased with the main branch. However, while addressing review comments today, I discovered that the code is incompatible with the Docker container and fails to compile. I wasn't able to find a Docker container matching the current code, so I merged with the main branch and pulled the latest Docker image. For some unknown reason, this PR now shows over 34,000 lines of code changes. I'm uncertain on how to resolve this issue.

Moonm3n commented:
> This pull request has encountered some issues. Earlier, it was unable to pass the CI checks, and based on suggestions from the StarRocks team, I rebased with the main branch. However, while addressing review comments today, I discovered that the code is incompatible with the Docker container and fails to compile. I wasn't able to find a Docker container matching the current code, so I merged with the main branch and pulled the latest Docker image. For some unknown reason, this PR now shows over 34,000 lines of code changes. I'm uncertain on how to resolve this issue.

Thanks to StarRocks classmates for their help, the problem has been solved.

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[356]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

Moonm3n committed:
retry CI

Signed-off-by: Moonm3n <saxonzhan@gmail.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AnalyticExpr.java|3|4|75.00%|[254]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyticAnalyzer.java|4|5|80.00%|[356]|
|:large_blue_circle:|com/starrocks/catalog/FunctionSet.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/transformer/WindowTransformer.java|3|3|100.00%|[]|

fzhedu reviewed (approved):
LGTM

packy92 commented on fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:
```diff
@@ -2116,6 +2116,7 @@ windowFunction
     : name = ROW_NUMBER '(' ')'
     | name = RANK '(' ')'
     | name = DENSE_RANK '(' ')'
+    | name = CUME_DIST '(' ')'
```
better add this word to `nonReserved` in `Starrocks.g4`. If not, sql like `select cume_dist from tbl` will fail to parse if `cume_dist` is a col name in tbl.

ChengyangDu commented:
Question: how does state.count update?

I just read the source code, and I cannot figure out how state.count updates. There is no related operation in update_batch_single_state_with_frame function.
so would you pls. provide some explanations. Great thx!
